test/k8sT/CustomCalls.go

```diff
@@ -0,0 +1,306 @@
+				// and the comparison on the counter value will
+				// fail. Could be adjusted in the future.
+				Failf("Direction not supported: %v", direction)
+			}
+
+			By("Updating call map with reference to custom program")
+
+			// TODO: other directions
+			cmd := fmt.Sprintf("bpftool map update pinned %scilium_calls_custom_%05d key %d 0 0 0 value pinned %s", bpffsDir, endpointId, direction, progPinPath)
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to update call map with reference to custom program ('%s'): %s", cmd, res.Stderr()))
+
+			By("Sending traffic between the pods")
+
+			cmd = helpers.Ping(serverIP)
+			res = kubectl.ExecPodCmd(helpers.DefaultNamespace, clientName, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to ping from %s to %s", "", ""))
+
+			By("Retrieving counter value")
+
+			cmd = fmt.Sprintf("bpftool -j map lookup pinned %s key %s | jq -r '.value[]'", mapPinPath, serverIdentity)
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to lookup byte counter value ('%s'): %s", cmd, res.Stderr()))
+
+			// Output from bpftool is on the form
+			// "0x11\n0x22\n0x33\n...0x88\n", convert it.
+			// Too bad bpftool returns hex that jq doesn't know how
+			// to convert yet.
+			var count uint64 = 0
+			byteString := strings.TrimSpace(res.Stdout())
+			bytes := strings.Split(byteString, "\n")
+			for i, b := range bytes {
+				dec, err := strconv.ParseUint(b, 0, 8)
+				Expect(err).ToNot(HaveOccurred(), fmt.Sprintf("Failed to convert byte counter value to an integer (%s from %q): %s", b, bytes, err))
+				count += dec << (i * 8)
+			}
+
+			By("Checking counter value")
+
+			expectedCount := 98 * helpers.PingCount
+			Expect(int(count)).To(Equal(expectedCount), fmt.Sprintf("Byte count (%d) differs from expected value (%d)", count, expectedCount))
+		}
+
+		cleanupByteCounter := func(endpointId int64, ciliumPod string, serverIdentity string, direction customCallDirection) {
+			// Clean up tail call map entry
+			cmd := fmt.Sprintf("bpftool map delete pinned %scilium_calls_custom_%05d key %d 0 0 0", bpffsDir, endpointId, direction)
+			res := kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to remove reference to byte counter program from tail call map ('%s'): %s", cmd, res.Stderr())
+			}
+
+			// Reset byte counter entry
+			cmd = fmt.Sprintf("bpftool map delete pinned %s key %s", mapPinPath, serverIdentity)
+			res = kubectl.ExecPodCmd(helpers.KubeSystemNamespace, ciliumPod, cmd)
+			if res.GetExitCode() != 0 {
+				log.Warningf("Failed to delete entry from byte counter map ('%s'): %s", cmd, res.Stderr())
+			}
+		}
+
+		It("Loads byte counter and get consistent values", func() {
+
+			// Deploy Cilium, enable tail calls to custom programs
+			deploymentManager.DeployCilium(map[string]string{
+				"customCalls.enabled": "true",
```

Comment:
Could we run the same with per-endpoint routes as well? Otherwise, one of the three IPv4 hook is not covered.

Reply:
Good point, thank you!